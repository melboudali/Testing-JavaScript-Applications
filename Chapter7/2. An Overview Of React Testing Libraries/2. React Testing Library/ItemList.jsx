import React from "react";

const ItemList = ({ items }) => {
	return (
		<ul>
			{Object.entries(items).map(([itemName, quantity]) => {
				return (
					<li key={itemName}>
						{itemName} - Quantity: {quantity}
					</li>
				);
			})}
		</ul>
	);
};
export default ItemList;
