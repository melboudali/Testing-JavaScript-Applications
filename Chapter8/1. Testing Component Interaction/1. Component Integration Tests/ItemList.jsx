import React from "react";

const ItemList = ({ items }) => {
	return (
		<ul>
			{Object.entries(items).map(([itemName, quantity]) => {
				return <li key={itemName}>{generateItemText(itemName, quantity)}</li>;
			})}
		</ul>
	);
};

export const generateItemText = (itemName, quantity) => {
	return `${itemName} - Quantity: ${quantity}`;
};

export default ItemList;
