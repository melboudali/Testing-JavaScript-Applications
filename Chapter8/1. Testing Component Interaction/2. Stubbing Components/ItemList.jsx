import React from "react";
import { Transition } from "react-spring";

const ItemList = ({ items }) => {
	const itemsList = Object.entries(items);
	return (
		<ul>
			<Transition
				items={itemsList}
				initial={null}
				keys={([itemName]) => itemName}
				from={{ fontSize: 0, opacity: 0 }}
				enter={{ fontSize: 18, opacity: 1 }}
				leave={{ fontSize: 0, opacity: 0 }}>
				{([itemName, quantity]) =>
					styleProps =>
						(
							<li key={itemName} style={styleProps}>
								{generateItemText(itemName, quantity)}
							</li>
						)}
			</Transition>
		</ul>
	);
};

export const generateItemText = (itemName, quantity) => {
	return `${itemName} - Quantity: ${quantity}`;
};

export default ItemList;
