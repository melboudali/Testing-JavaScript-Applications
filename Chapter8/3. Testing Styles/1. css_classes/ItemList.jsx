import React from "react";

const ItemList = ({ items }) => {
	const itemsList = Object.entries(items);
	return (
		<ul>
			{itemsList.map(([itemName, quantity]) => (
				<li key={itemName} className={quantity < 5 ? "almost-out-of-stock" : null}>
					{generateItemText(itemName, quantity)}
				</li>
			))}
		</ul>
	);
};

export const generateItemText = (itemName, quantity) => {
	return `${itemName} - Quantity: ${quantity}`;
};

export default ItemList;
