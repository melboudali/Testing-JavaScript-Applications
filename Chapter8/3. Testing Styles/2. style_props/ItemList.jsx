import React from "react";

const almostOutOfStock = { fontWeight: "bold", color: "red" };

const ItemList = ({ items }) => {
	const itemsList = Object.entries(items);
	return (
		<ul>
			{itemsList.map(([itemName, quantity]) => (
				<li key={itemName} style={quantity < 5 ? almostOutOfStock : null}>
					{generateItemText(itemName, quantity)}
				</li>
			))}
		</ul>
	);
};

export const generateItemText = (itemName, quantity) => {
	return `${itemName} - Quantity: ${quantity}`;
};

export default ItemList;
