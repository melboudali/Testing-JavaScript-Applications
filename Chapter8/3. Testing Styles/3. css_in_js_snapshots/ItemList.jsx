import React from "react";
import styled, { css, keyframes } from "styled-components";

const pulsate = keyframes`
	0% { opacity: .3; }
	50% { opacity: 1; }
	100% { opacity: .3; }
`;

const ListItem = styled.li`
	font-weight: ${({ quantity }) => (quantity < 5 ? "bold" : "normal")};
	color: ${({ quantity }) => (quantity < 5 ? "red" : "black")};
	${({ quantity }) =>
		quantity < 5 &&
		css`
			animation: ${pulsate} 2s infinite;
		`}
`;

const ItemList = ({ items }) => {
	const itemsList = Object.entries(items);
	return (
		<ul>
			{itemsList.map(([itemName, quantity]) => (
				<ListItem key={itemName} quantity={quantity}>
					{generateItemText(itemName, quantity)}
				</ListItem>
			))}
		</ul>
	);
};

export const generateItemText = (itemName, quantity) => {
	return `${itemName} - Quantity: ${quantity}`;
};

export default ItemList;
