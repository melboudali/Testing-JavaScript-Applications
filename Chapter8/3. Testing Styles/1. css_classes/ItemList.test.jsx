import React from "react";
import ItemList, { generateItemText } from "./ItemList.jsx";
import { render } from "@testing-library/react";

describe("ItemList Component", () => {
	test("check items list", () => {
		const Items = { cheesecake: 2, croissant: 5, macaroon: 96 };
		const { getByText } = render(<ItemList items={Items} />);

		const listElement = document.querySelector("ul");
		expect(listElement.childElementCount).toBe(3);
		expect(getByText(generateItemText("cheesecake", 2))).toBeInTheDocument();
		expect(getByText(generateItemText("croissant", 5))).toBeInTheDocument();
		expect(getByText(generateItemText("macaroon", 96))).toBeInTheDocument();
	});

	test("highlighting items that are almost out of stock", () => {
		const Items = { cheesecake: 2, croissant: 5, macaroon: 96 };

		const { getByText } = render(<ItemList items={Items} />);
		const cheesecakeItem = getByText(generateItemText("cheesecake", 2));
		expect(cheesecakeItem).toHaveClass("almost-out-of-stock");
	});

	test("highlighting items that are almost out of stock with snapshot", () => {
		const Items = { cheesecake: 2, croissant: 5, macaroon: 96 };

		const { getByText } = render(<ItemList items={Items} />);
		const cheesecakeItem = getByText(generateItemText("cheesecake", 2));
		const macaroon = getByText(generateItemText("macaroon", 96));
		expect(cheesecakeItem).toMatchSnapshot();
		expect(macaroon).toMatchSnapshot();
	});
});

describe("generateItemText", () => {
	test("generating an item's text", () => {
		expect(generateItemText("cheesecake", 3)).toBe("cheesecake - Quantity: 3");
		expect(generateItemText("apple pie", 22)).toBe("apple pie - Quantity: 22");
	});
});
