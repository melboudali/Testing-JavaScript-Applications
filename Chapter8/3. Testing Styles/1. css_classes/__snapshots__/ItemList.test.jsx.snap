// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList Component highlighting items that are almost out of stock with snapshot 1`] = `
<li
  class="almost-out-of-stock"
>
  cheesecake - Quantity: 2
</li>
`;

exports[`ItemList Component highlighting items that are almost out of stock with snapshot 2`] = `
<li>
  macaroon - Quantity: 96
</li>
`;
